src/state/workload.rs

```diff
@@ -656,13 +679,34 @@ impl WorkloadStore {
         self.by_addr.get(addr).cloned()
     }
 
+    /// Finds the workload by workload information, as an arc.
+    pub fn find_by_info(&self, wl: &WorkloadInfo) -> Option<Arc<Workload>> {
+        // We do not have an index directly on the full workload info, but we can narrow it down
+        // to only workloads on the same node with the same identity -- a tiny set to iterate over
+        self.node_local_by_identity
+            .get(&WorkloadIdentity {
+                namespace: strng::new(&wl.namespace),
+                service_account: strng::new(&wl.service_account),
+            })?
+            .iter()
+            .find_map(|uid| self.by_uid.get(uid).cloned())
```

Comment:
It seems to me `find_map`'s function is missing a check to ensure that the result of `by_uid.get(uid)` actually matches wl, not just shares a node and identity and was first in the HashSet.iter().

Reply:
This probably should be the invert of `identity/manager.rs`'s `matches` - e.g. we just add another check in the `find_map` closure to assert that `wl.Name == xds_wl.Name` and it's As Good As that existing check

Reply:
(or we just use that func, actually)

Reply:
great call :+1: updated